fix(profile): Update redirection if not logged in or different user

c-koehler opened the pull request:
fixes https://github.com/openshiftio/openshift.io/issues/871

fabric8cd commented:
@c-koehler fabric8/fabric8-ui:SNAPSHOT-PR-2271-2 fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2271-fabric8-ui.badger.fabric8.io

dgutride commented on src/app/profile/update/update-guard.component.ts:
```diff
@@ -0,0 +1,59 @@
+      // TODO: Right now it's a hardcoded link, so can't grab user (like params['user']).
+      this.router.url.split('/').forEach(urlToken => {
+        if (urlToken === this.username) {
+          this.differentUsername = false;
```
You should change this function to return a boolean instead of using a temporary variable that is only used in the line immediately after this is called on line 27

dgutride commented on src/app/profile/update/update-guard.component.ts:
```diff
@@ -0,0 +1,59 @@
+    this.activateObservable = Observable.create(observer => {
+      this.subscription = this.userService.loggedInUser.subscribe(user => {
+        this.handleUserNotification(user);
+        if (this.differentUsername) {
```
You can simplify this by removing this.differentUsername and just returning a boolean from the handleUserNotification maybe renaming that function to indicate what it's actually checking.

c-koehler commented:
@dgutride It has been updated to no longer use temporary variables, and also solved the `TODO` issue. I refactored it so that this service is now accessible to anyone who wants this particular guard, and added in some missing things that were present in other guards from the `shared/` folder. Hopefully that also solves the other comment left as well.

fabric8cd commented:
@c-koehler fabric8/fabric8-ui:SNAPSHOT-PR-2271-3 fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2271-fabric8-ui.badger.fabric8.io

joshuawilson commented:
Is there a test for the service?

c-koehler commented:
@joshuawilson Tests now available.

c-koehler commented:
@dgutride When you get a chance can you let me know if this is okay?

fabric8cd commented:
@c-koehler fabric8/fabric8-ui:SNAPSHOT-PR-2271-8 fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2271-fabric8-ui.badger.fabric8.io

dgutride commented on src/app/shared/owner-guard.service.ts:
```diff
@@ -0,0 +1,82 @@
+    if (this.authService.isLoggedIn()) {
+      return Observable.create(observer => {
+        this.subscription = this.userService.loggedInUser.subscribe(user => {
+          observer.next(this.userMatchesOwnerFromRoute(user));
```
I added a helper function to the context service that checks if a user is viewing his own stuff (see viewingOwnContext() in shared/context.service.ts).  It will simplify this code to use that.

c-koehler commented:
@dgutride This cleans up the code a lot, it's pretty nice!

One thing I noticed is that a user can access their own profile when navigating around, but they cannot directly go to their profile (ex: going to http://localhost:3000/chkoehle/_update will not work, I have to specifically navigate to it when I'm at my profile at http://localhost:3000/chkoehle), just checking with you to make sure that it makes sense, as doing some debugging showed that `viewingOwnContext()` returns `false` if not doing it from the link profile provided.

fabric8cd commented:
@c-koehler fabric8/fabric8-ui:SNAPSHOT-PR-2271-10 fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2271-fabric8-ui.badger.fabric8.io
